### PR TITLE
Clean up a variety of leaks

### PR DESCRIPTION
**Pull request by raskchanky**
The aim of this PR is to plug as many of the thread/file descriptor/pipe leaks as we reasonably can, and generally reduce thread usage if possible.

The number of HTTP gateway threads has been capped at 2. I could be way off base here, but the supervisor's HTTP gateway doesn't seem like a high traffic server. The default number of threads for the HTTP gateway is equal to the number of logical CPU cores for the machine it's running on. I capped it at 2, thinking that given its low usage, that's probably enough.

The thread that watches user.toml for changes now uses `try_recv` instead of `recv`, meaning that 2 out of the 3 threads now shutdown properly when a service is unloaded, without needing any extra shenanigans regarding `touch`ing files.

The final thread is part of the `DebouncedEvent` API of the notify crate. After much discussion, investigation, wringing of hands and gnashing of teeth, it was decided that converting the [file watcher](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/file_watcher.rs) abstraction and all of its associated files to use the `RawEvent` API of the notify crate was too much of an investment to make at this time. Therefore, we're still leaking 1 thread for watching user.toml files. From the sound of various issues on the notify crate's repo, the author is preparing a new release that should theoretically solve a lot of the threading and performance issues with the current `DebouncedEvent` API.

Finally, the service updater thread now dies correctly when a service is removed via `hab svc unload`.

All in all, while this doesn't solve the problem 100%, it's a dramatic improvement over the current state of affairs. I will open an issue to track the final thread leak.

Closes https://github.com/habitat-sh/habitat/issues/5416

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by raskchanky** [comment]
@jaym This might be of interest to you, as you opened the original issue.

**Comment by stevendanna** [comment]
I think this may also fix:

https://github.com/habitat-sh/habitat/issues/5734
https://github.com/habitat-sh/habitat/issues/5734

And quite possibly significantly improve:

https://github.com/habitat-sh/habitat/issues/5038

**Comment by jaym** [comment]
I suspect #4815 might be fixed with this

**Comment by stevendanna** [comment]
Thanks for working on this!  I have a feeling this is going to solve a lot of other "mystery" problems for us as well.

File: components/sup/src/manager/service_updater.rs [hunk 1]
```diff
@@ -88,8 +89,9 @@ impl ServiceUpdater {
                 self.states
                     .entry(service.service_group.clone())
                     .or_insert_with(|| {
-                        let rx = Worker::new(service).start(&service.service_group, None);
-                        UpdaterState::AtOnce(rx)
+                        let (kill_tx, kill_rx) = channel();
```
Comment (christophermaier) [inline components/sup/src/manager/service_updater.rs]: Not for this PR, but it seems like there's an abstraction hiding in here that we might be able to extract... some kind of "kill switch". There's two instances of this in this PR alone, I've got a new one I added in my current work, and there will likely be instances in the future where this pattern could come in handy.

Worth pondering, at any rate 🤔

File: components/sup/src/manager/service_updater.rs [hunk 2]
```diff
@@ -102,8 +104,46 @@ impl ServiceUpdater {
...
+                    FollowerState::Updating(_rx, kill_tx) => {
+                        if kill_tx.send(true).is_err() {
+                            debug!("Tried to kill the updater thread but it's already dead.");
+                        }
+                    }
+                    FollowerState::Waiting => {}
+                },
+                _ => {}
+            },
+            None => {}
```
Comment (christophermaier) [inline components/sup/src/manager/service_updater.rs]: This is handled above on line 111... might as well consolidate the entire body of this function into one big `match` statement.

File: components/sup/src/manager/service_updater.rs [hunk 3]
```diff
@@ -384,67 +445,100 @@ impl Worker {
...
+        ident: PackageIdent,
+        kill_rx: Receiver<bool>,
+    ) {
         // Fairly certain that this only gets called in a rolling update
         // scenario, where `ident` is always a fully-qualified identifier
         outputln!("Updating from {} to {}", self.current, ident);
         let install_source = (ident, *PackageTarget::active_target()).into();
-        loop {
-            let next_time = self.next_period_start();
+        let mut next_time = None::<SteadyTime>;
```
Comment (christophermaier) [inline components/sup/src/manager/service_updater.rs]: I think if you initially set `next_time` to "now", you'll get the same ultimate effect, without the verbosity of using an `Option` here.

```suggestion
        let mut next_time = SteadyTime::now();
```

File: components/sup/src/manager/service_updater.rs [hunk 4]
```diff
@@ -384,67 +445,100 @@ impl Worker {
...
+            }
         }
     }
 
     /// Continually poll for a new version of a package, installing it
     /// when found.
-    fn run_poll(&mut self, sender: SyncSender<PackageInstall>) {
+    fn run_poll(&mut self, sender: Sender<PackageInstall>, kill_rx: Receiver<bool>) {
         let install_source = (self.spec_ident.clone(), *PackageTarget::active_target()).into();
+        let mut next_time = None::<SteadyTime>;
```
Comment (christophermaier) [inline components/sup/src/manager/service_updater.rs]: Same suggestion here

File: components/sup/src/http_gateway.rs [hunk 5]
```diff
@@ -161,7 +161,8 @@ impl Server {
             let bind = server::new(move || {
                 let app_state = AppState::new(gateway_state.clone());
                 App::with_state(app_state).configure(routes)
-            }).bind(listen_addr.to_string());
+            }).workers(2)
+            .bind(listen_addr.to_string());
```
[the remark this answers is not in the document]
Comment (baumanj) [reply components/sup/src/http_gateway.rs]: I'm always in favor of the undocumented env var for tuning magic numbers. It'll probably never be needed, but in the situation is it, we'll be glad we have it and it's low risk/complexity.

File: components/sup/src/manager/service_updater.rs [hunk 6]
```diff
@@ -384,67 +445,100 @@ impl Worker {
...
-                    sender.send(package).expect("Main thread has gone away!");
+        loop {
+            match kill_rx.try_recv() {
+                Ok(_) => {
+                    info!("Received some data on the kill channel. Letting this thread die.");
+                    break;
+                }
+                Err(TryRecvError::Empty) => {}
+                Err(TryRecvError::Disconnected) => {
+                    info!("Service updater has gone away, yikes!");
```
Comment (baumanj) [inline components/sup/src/manager/service_updater.rs]: Given that it's "yikes"-worthy, perhaps a `warn!` (or even `error!`)?

File: components/sup/src/manager/service_updater.rs [hunk 7]
```diff
@@ -384,67 +445,100 @@ impl Worker {
...
         loop {
-            let next_time = self.next_period_start();
+            match kill_rx.try_recv() {
+                Ok(_) => {
+                    info!("Received some data on the kill channel. Letting this thread die.");
+                    break;
+                }
+                Err(TryRecvError::Empty) => {}
+                Err(TryRecvError::Disconnected) => {
+                    info!("Service updater has gone away, yikes!");
```
Comment (baumanj) [inline components/sup/src/manager/service_updater.rs]: Ditto

File: components/sup/src/manager/service_updater.rs [hunk 8]
```diff
@@ -37,7 +38,7 @@ const DEFAULT_FREQUENCY: i64 = MIN_ALLOWED_FREQUENCY;
 type UpdaterStateList = HashMap<ServiceGroup, UpdaterState>;
 
 enum UpdaterState {
-    AtOnce(Receiver<PackageInstall>),
+    AtOnce(Receiver<PackageInstall>, Sender<bool>),
```
Comment (baumanj) [inline components/sup/src/manager/service_updater.rs]: It's not immediately obvious to me what the `bool` here represents. Perhaps an `enum` would make it clearer?

This goes for the other `Sender<bool>`s as well.

Comment (raskchanky) [reply components/sup/src/manager/service_updater.rs]: I'm all for saving bytes.  And yes, @christophermaier is right that I just wanted to send a signal.  The value is irrelevant.  I'll change it to `()`

File: components/sup/src/manager/service_updater.rs (hunk 2 above)
Comment (raskchanky) [reply components/sup/src/manager/service_updater.rs]: Good catch

File: components/sup/src/manager/service_updater.rs (hunk 3 above)
Comment (raskchanky) [reply components/sup/src/manager/service_updater.rs]: Ah, good call.

File: components/sup/src/manager/service_updater.rs (hunk 6 above)
Comment (raskchanky) [reply components/sup/src/manager/service_updater.rs]: Will do

File: components/sup/src/manager/file_watcher.rs [hunk 9]
```diff
@@ -3184,6 +3184,11 @@ mod tests {
 
         fn spin_watcher(&self, setup: &mut WatcherSetup, iterations: u32) {
             let mut iteration = 0;
+
+            // After switching single_iteration() from recv() to try_recv(), this sleep is required
+            // for these tests to pass.
+            ::std::thread::sleep(Duration::from_secs(3));
```
Comment (christophermaier) [inline components/sup/src/manager/file_watcher.rs]: tenor-34493659

**Pull request merged**